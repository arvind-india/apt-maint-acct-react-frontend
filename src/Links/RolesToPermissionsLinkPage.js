import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
//import { Table, Alert, UncontrolledAlert } from 'reactstrap'
//import { Router, Route } from 'react-router-dom'

import {
    Input,
    Label
} from 'reactstrap'

import {
  MdHome,
  //MdAdd,
  //MdVisibility,
  //MdDelete,
  //MdContentCut,
  //MdAttachment,
  MdThumbUp,
  MdThumbDown
} from 'react-icons/lib/md' // material design icons

import {
  Button
} from 'reactstrap'

//import { history } from '../_helpers'
import { FlashMessage } from '../_components'
import { roleActions,  permissionActions } from '../_actions'


//let url = '/rolesToPermissions'

class RolesToPermissionsLinkPage extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      selectedOptionInLeftList: '',
      selectedOptionsInDList: [],
      selectedOptionsInAList: []
    }
    this.handleChangeInLeftList = this.handleChangeInLeftList.bind(this)
    this.handleChangeInDetachedList = this.handleChangeInDetachedList.bind(this)
    this.attachItems = this.attachItems.bind(this)
    this.handleChangeInAttachedList = this.handleChangeInAttachedList.bind(this)
    this.detachItems = this.detachItems.bind(this)
  }

  componentDidMount() {
    this.props.dispatch(roleActions.getAll())
    this.props.dispatch(permissionActions.getAll())
  }

  render() {
    console.log('Props in RolesToPermissionsLinkPage: ', this.props)
    const { alert } = this.props
    return (
      <div>
        <h3>Roles To Permissions Link</h3>
        {alert.message && <FlashMessage text={alert.message} delay={5000}/>}
        {this.show()}
      </div>
    )
  }

  show() {
    return <div className="list-to-list">
      {this.showLeftList()}
      {this.showAttachedList()}
      {this.showDetachedList()}
    </div>
  }

  showLeftList() {
    const { roles } = this.props
//    console.log('Roles: ', roles)
    return <div>
      <Label for="leftItem" className="llabel">Select a Role</Label>
      <Input
        type="select"
        name="leftItem"
        id="leftItem"
        size="20"
        className="lselect"
        onChange={this.handleChangeInLeftList}
      >
      { roles.items && roles.items.map(each =>
        <option value={each.id} title={each.description} key={each.id}
          >{each.name}</option>)
      }
      </Input>
      <Button color="link" className="homeButton"><Link to="/home"><MdHome/> Home</Link></Button>
    </div>
  }

  handleChangeInLeftList(event) {
    const { value } = event.target
//    console.log('Left list name: ', name); console.log('Left list value: ', value)
    this.setState({
      selectedOptionInLeftList: value
    })
    this.props.dispatch(roleActions.getMyPermissions(value))
  }

  showAttachedList() {
    const { rolesToPermissions, authzn } = this.props
    const { selectedOptionInLeftList, selectedOptionsInAList } = this.state
// console.log('rolesToPermissions: ', rolesToPermissions)
    return <div>
      <Label
        for="attachedItems"
        className="alabel"
      >Granted Permissions</Label>
      <Input
        type="select"
        name="attachedItems"
        id="attachedItems"
        size="20"
        className="aselect"
        onChange={this.handleChangeInAttachedList}
        multiple
      >
      { selectedOptionInLeftList &&
        rolesToPermissions.items &&
        rolesToPermissions.items.map(each =>
        <option value={each.id} title={each.description} key={each.id}
          >{each.operations} on {each.resource} {each.condition?' (restricted)':''}</option>)
      }
      </Input>
      <Button
        color="danger"
        className="dbutton"
        onClick={this.detachItems}
        disabled={selectedOptionsInAList.length === 0}
        hidden={!authzn.allowsEdit}
      ><MdThumbDown/> Revoke</Button>
    </div>
  }
  handleChangeInAttachedList(event) {
    const { selectedOptions } = event.target
    this.setState({
      selectedOptionsInAList: Array.from(selectedOptions)
    })
  }
  detachItems() {
    const { dispatch, rolesToPermissions } = this.props
    const { selectedOptionsInAList, selectedOptionInLeftList } = this.state
//    console.log('selected options: ', selectedOptionsInAList)
    let id = selectedOptionInLeftList
    let ids_toBeRemoved = selectedOptionsInAList.map(e => e.value)
    let ids = rolesToPermissions.items.map(e => e.id.toString())
    let ids_toBeRetained = ids.filter(each => !ids_toBeRemoved.includes(each))
//    console.log('role id: ', id); console.log('all ids: ', ids);
//    console.log('toBeRemoved: ',ids_toBeRemoved);console.log('toBeRetained: ', ids_toBeRetained)
    dispatch(roleActions.updateMyPermissions(id, ids_toBeRetained))
    this.setState({ selectedOptionsInAList: [] })
  }
  showDetachedList() {
    const { rolesToPermissions, permissions, authzn } = this.props
    const { selectedOptionInLeftList, selectedOptionsInDList } = this.state
//    console.log('Available permissions: ', permissions)
    let available = [];
    let grantedIDs = []
    if(rolesToPermissions.items) {
      grantedIDs = rolesToPermissions.items.map(each => each.id)
    }
    if(permissions.items && selectedOptionInLeftList) {
      available = grantedIDs.length ?
        permissions.items.filter(each => !grantedIDs.includes(each.id)) :
        permissions.items
    }
    return <div>
      <Label
        for="detachedItems"
        className="dlabel"
      >Available Permissions</Label>
      <Input
        type="select"
        name="detachedItems"
        id="detachedItems"
        size="20"
        className="dselect"
        onChange={this.handleChangeInDetachedList}
        multiple
      >
      {
        available.map(each =>
        <option value={each.id} title={each.description} key={each.id}
          >{each.operations} on {each.resource} {each.condition?' (restricted)':''}</option>)
      }
      </Input>
      <Button
        color="success"
        className="abutton"
        onClick={this.attachItems}
        disabled={selectedOptionsInDList.length === 0}
        hidden={!authzn.allowsEdit}
      ><MdThumbUp/> Grant</Button>
    </div>
  }

  handleChangeInDetachedList(event){
    const { selectedOptions } = event.target
    this.setState({
      selectedOptionsInDList: Array.from(selectedOptions)
    })
  }
  attachItems() {
    const { dispatch, rolesToPermissions } = this.props
    const { selectedOptionsInDList, selectedOptionInLeftList } = this.state
//    console.log('selected options: ', selectedOptionsInDList)
    let id = selectedOptionInLeftList
    let ids = rolesToPermissions.items.map(e => e.id)
    let ids_toBeAdded = selectedOptionsInDList.map(e => e.value)
    let new_ids = ids.concat(ids_toBeAdded)
    console.log('role id: ', id);
    console.log('existing perms ids: ', ids);
    console.log('permissions ids to be added: ', ids_toBeAdded);
    console.log('new_ids: ', new_ids)
    dispatch(roleActions.updateMyPermissions(id, new_ids))
    this.setState({ selectedOptionsInDList: [] })
  }
}


function mapStateToProps(state) {
  const { authentication, alert, roles, rolesToPermissions, permissions, authorizations } = state
  const { user } = authentication
  const authzn = authorizations.roles-permissions
  return {
    alert,
    user,
    roles,
    rolesToPermissions,
    permissions,
    authzn
  }
}

const connectedLinkPage = connect(mapStateToProps)(RolesToPermissionsLinkPage)
export { connectedLinkPage as RolesToPermissionsLinkPage }
