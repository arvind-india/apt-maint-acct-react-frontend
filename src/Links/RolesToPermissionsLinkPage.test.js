// RolesToPermissionsLinkPage.test.js0]

import React from 'react'
import { shallow } from 'enzyme'

import { RolesToPermissionsLink } from './RolesToPermissionsLinkPage'

const mockRoles = [
  { id: 1, name: 'guest', description: 'guest role' },
  { id: 2, name: 'tenant', description: 'tenant role' }
]

const mockPermissions = [
  { id: 1, operations: 'R', condition: '', resource: 'accounts', description: 'Read only operations on Accounts module' },
  { id: 2, operations: 'RU', condition: 'userOwnRecord', resource: 'users', description: 'Read and Update own records in Users module' }
]

const mockProps = [
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    roles: { items: mockRoles },
    rolesToPermissions: { items: [] },
    permissions: { items: mockPermissions },
    getAllRoles: jest.fn(),
    getAllPermissions: jest.fn(),
    getMyPermissions: jest.fn(),
    updateMyPermissions: jest.fn()
  }
]

const component = shallow(<RolesToPermissionsLink {...mockProps[0]} />)

describe('RolesToPermissionsLinkPage params test', () => {
  it('should render RolesPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
  it('should have model state', () => {
    expect(component.state('selectedOptionInLeftList')).toBeDefined()
  })
  it('should have model state', () => {
    expect(component.state('selectedOptionsInDList')).toBeDefined()
  })
  it('should have model state', () => {
    expect(component.state('selectedOptionsInAList')).toBeDefined()
  })
})


const selectedOptions = [mockPermissions[0].id, mockPermissions[1].id]

describe('RolesToPermissionsLinkPage test: select a role and select two permissions', () => {

  component.find('#leftItem').simulate('change', {target: {name:'id', value: mockRoles[0].id}})

  it('should check options in Left List', () => {
    expect(component.state('selectedOptionInLeftList')).toEqual(mockRoles[0].id)
  })

  component
    .find('#detachedItems')
    .simulate('change', { target: {name: 'id', selectedOptions: selectedOptions} })

  it('should check options in detached list', () => {
    expect(component.state('selectedOptionsInDList'))
      .toEqual(selectedOptions)
  })

  component
    .find('#attachedItems')
    .simulate('change', { target: {name: 'id', selectedOptions: selectedOptions} })

  it('should have two options in the attached list', () => {
    expect(component.state('selectedOptionsInAList'))
      .toEqual(selectedOptions)
  })

})


const component2 = shallow(<RolesToPermissionsLink {...mockProps[0]} />)

describe('RolesToPermissionsLinkPage test: Attach selected items', () => {

  component2
    .find('#leftItem')
    .simulate('change', {target: {name:'id', value: mockRoles[0].id} })

  component2
    .find('#detachedItems')
    .simulate('change', {target: {name: 'id', selectedOptions: selectedOptions} })

  component2
    .find('Button.abutton')
    .simulate('click')

  it('should now remove two options from detached list', () => {
    expect(component2.state('selectedOptionsInDList')).toEqual([])
  })

})



const component3 = shallow(<RolesToPermissionsLink {...mockProps[0]} />)

describe('RolesToPermissionsLinkPage test: Detach selected items', () => {

  component3
    .find('#leftItem')
    .simulate('change', { target: {name:'id', value: mockRoles[0].id} })

  component3
    .find('#attachedItems')
    .simulate('change', { target: {name: 'id', selectedOptions: selectedOptions} })

  component3
    .find('Button.dbutton')
    .simulate('click')

  it('should now remove two options from attached list', () => {
    expect(component3.state('selectedOptionsInAList')).toEqual([])
  })

})
