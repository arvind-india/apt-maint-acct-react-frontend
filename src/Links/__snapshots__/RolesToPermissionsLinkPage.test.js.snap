// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RolesToPermissionsLinkPage params test should render initial layout 1`] = `
Array [
  <div>
    <h3>
        Roles To Permissions Link
    </h3>
    <FlashMessage
        delay={4000}
        text="testingAlert"
    />
    <div
        className="list-to-list"
    >
        <Label
            className="llabel"
            for="leftItem"
            tag="label"
            widths={
                Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
            }
        >
            Select a Role
        </Label>
        <Input
            className="lselect"
            id="leftItem"
            name="leftItem"
            onChange={[Function]}
            size="20"
            tag="p"
            type="select"
        >
            <option
                title="guest role"
                value={1}
            >
                guest
            </option>
            <option
                title="tenant role"
                value={2}
            >
                tenant
            </option>
        </Input>
        <div
            class="homeButton"
        >
            <Button
                color="link"
                tag="button"
            >
                <Link
                    replace={false}
                    to="/home"
                >
                    <MdHome />
                     Home
                </Link>
            </Button>
        </div>
        <Label
            className="alabel"
            for="attachedItems"
            tag="label"
            widths={
                Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
            }
        >
            Granted Permissions
        </Label>
        <Input
            className="aselect"
            id="attachedItems"
            multiple={true}
            name="attachedItems"
            onChange={[Function]}
            size="20"
            tag="p"
            type="select"
        >
            
        </Input>
        <Button
            className="dbutton"
            color="danger"
            disabled={false}
            hidden={false}
            onClick={[Function]}
            tag="button"
            title="Revoke selected roles"
        >
            <MdThumbDown />
             Revoke
        </Button>
        <Label
            className="dlabel"
            for="detachedItems"
            tag="label"
            widths={
                Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
            }
        >
            Available Permissions
        </Label>
        <Input
            className="dselect"
            id="detachedItems"
            multiple={true}
            name="detachedItems"
            onChange={[Function]}
            size="20"
            tag="p"
            type="select"
        >
            <option
                title="Read only operations on Accounts module"
                value={1}
            >
                accounts
                 
                
                 with 
                R
            </option>
            <option
                title="Read and Update own records in Users module"
                value={2}
            >
                users
                 
                 (restricted)
                 with 
                RU
            </option>
        </Input>
        <Button
            className="abutton"
            color="success"
            disabled={false}
            hidden={false}
            onClick={[Function]}
            tag="button"
            title="Grant selected roles"
        >
            <MdThumbUp />
             Grant
        </Button>
    </div>
</div>,
]
`;
